lib/checkother.cpp

```diff
@@ -1391,6 +1391,16 @@ void CheckOther::checkConstVariable()
         // Skip if address is taken
         if (Token::findmatch(var->nameToken(), "& %varid%", scope->bodyEnd, var->declarationId()))
             continue;
+        // Skip if another non-const variable is initialized with this variable
+        {
+            auto match = Token::findmatch(var->nameToken(), "%type% & %name% = %varid%", scope->bodyEnd, var->declarationId());
+            if (match)
+                if (!Token::simpleMatch(match->previous(), "const"))
+                    continue;
+        }
+        // Skip if we ever dynamic_cast/static_cast this variable to a non-const type
+        if (Token::findmatch(var->nameToken(), "dynamic_cast|static_cast < %type% & > ( %varid% )", scope->bodyEnd, var->declarationId()))
```

Comment:
How would this handle `dynamic_cast<typename T::U&>(x)` or `dynamic_cast<T*>(&x)`?

Reply:
There's a case already for taking the address of the variable which I would expect to capture ```dynamic_cast<T*>(&x)``` but I can add a test for that regardless.

It probably won't capture ```dynamic_cast<typename T::U&>(x)``` currently. Is there a way to add optional tokens to the match string or should I just test against two different strings: one with and one without typename?

Reply:
I've added tests for both of these cases and modified the pattern to optionally include typename

Reply:
I wonder if it would be better to traverse the ast, but that requires a little more work because we dont have any ast matchers.

Reply:
I think I prefer it this way but I could be convinced.